testmodule/testCustomErrors.nim

```diff
@@ -0,0 +1,137 @@
+      fail()
+    except ErrorWithStaticStruct as error:
+      check error.arguments.one == (1.u256, 2.u256)
+      check error.arguments.two == (3.u256, 4.u256)
+
+  test "handles error with dynamic struct arguments":
+    proc revertsErrorWithDynamicStruct(contract: TestCustomErrors)
+      {.contract, pure, errors:[ErrorWithDynamicStruct].}
+
+    try:
+      await contract.revertsErrorWithDynamicStruct()
+      fail()
+    except ErrorWithDynamicStruct as error:
+      check error.arguments.one == ("1", 2.u256)
+      check error.arguments.two == ("3", 4.u256)
+
+  test "handles error with dynamic and static struct arguments":
+    proc revertsErrorWithDynamicAndStaticStruct(contract: TestCustomErrors)
+      {.contract, pure, errors:[ErrorWithDynamicAndStaticStruct].}
+
+    try:
+      await contract.revertsErrorWithDynamicAndStaticStruct()
+      fail()
+    except ErrorWithDynamicAndStaticStruct as error:
+      check error.arguments.one == ("1", 2.u256)
+      check error.arguments.two == (3.u256, 4.u256)
+
+  test "handles gas estimation errors":
+    proc revertsTransaction(contract: TestCustomErrors)
+      {.contract, errors:[ErrorWithArguments].}
+
+    let contract = contract.connect(provider.getSigner())
+    try:
+      await contract.revertsTransaction()
+      fail()
+    except ErrorWithArguments as error:
+      check error.arguments.one == 1.u256
+      check error.arguments.two == true
+
+  test "handles transaction submission errors":
+    proc revertsTransaction(contract: TestCustomErrors)
+      {.contract, errors:[ErrorWithArguments].}
+
+     # skip gas estimation
+    let overrides = TransactionOverrides(gasLimit: some 1000000.u256)
+
+    let contract = contract.connect(provider.getSigner())
+    try:
+      await contract.revertsTransaction(overrides = overrides)
+      fail()
+    except ErrorWithArguments as error:
+      check error.arguments.one == 1.u256
+      check error.arguments.two == true
+
+  test "handles transaction confirmation errors":
+    proc revertsTransaction(contract: TestCustomErrors): ?TransactionResponse
+      {.contract, errors:[ErrorWithArguments].}
+
+     # skip gas estimation
+    let overrides = TransactionOverrides(gasLimit: some 1000000.u256)
+
+    # ensure that transaction is not immediately checked by hardhat
+    discard await provider.send("evm_setAutomine", @[%false])
+
+    let contract = contract.connect(provider.getSigner())
+    try:
+      let future = contract.revertsTransaction(overrides = overrides).confirm(0)
+      await sleepAsync(100.millis) # wait for transaction to be submitted
+      discard await provider.send("evm_mine", @[]) # mine the transaction
+      discard await future # wait for confirmation
+      fail()
+    except ErrorWithArguments as error:
+      check error.arguments.one == 1.u256
+      check error.arguments.two == true
+
+    # re-enable auto mining
+    discard await provider.send("evm_setAutomine", @[%true])
```

Comment:
Maybe it would make sense to add a test that tests for multiple error types?

```nim
test "handles multiple errors":
    proc revertsMultipleErrors(contract: TestCustomErrors, simple: bool)
      {.contract, errors:[SimpleError, ErrorWithArguments].}

     # skip gas estimation
    let overrides = TransactionOverrides(gasLimit: some 1000000.u256)

    let contract = contract.connect(provider.getSigner())
    expect SimpleError:
      await contract.revertsMultipleErrors(true, overrides = overrides)
    expect ErrorWithArguments:
      await contract.revertsMultipleErrors(false, overrides = overrides)
```
`TestCustomErrors.sol`:
```solidity
function revertsMultipleErrors(bool simple) public pure {
    if (simple) {
      revert SimpleError();
    }
    revert ErrorWithArguments(1, false);
  }
```

Reply:
Good addition! I added the suggested test.